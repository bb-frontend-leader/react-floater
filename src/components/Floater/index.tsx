import { cloneElement, CSSProperties, isValidElement, memo, ReactNode, Ref, useMemo } from 'react';
import { PlainObject } from '@gilbarbara/types';

import { STATUS } from '../../literals';
import { CloseFunction, FloaterComponent, Statuses, Styles } from '../../types';

import Arrow from './Arrow';
import Container from './Container';

interface Props {
  arrow?: ReactNode;
  arrowRef: Ref<HTMLSpanElement>;
  component?: FloaterComponent;
  content?: ReactNode;
  floaterRef: Ref<HTMLDivElement>;
  footer?: ReactNode;
  hideArrow: boolean;
  id: string;
  onClick: CloseFunction;
  placement: string;
  positionWrapper: boolean;
  showCloseButton?: boolean;
  status: Statuses;
  styles: Styles;
  title?: ReactNode;
}

function Floater(props: Props) {
  const {
    component,
    content,
    floaterRef,
    hideArrow,
    id,
    onClick: closeFn,
    placement,
    status,
    styles,
  } = props;

  const style = useMemo(() => {
    const {
      arrow: { length },
      floater,
      floaterCentered,
      floaterClosing,
      floaterOpening,
      floaterWithComponent,
    } = styles;
    let element: CSSProperties = { ...floater };

    if (!hideArrow) {
      if (placement.startsWith('top')) {
        element.padding = `0 0 ${length}px`;
      } else if (placement.startsWith('bottom')) {
        element.padding = `${length}px 0 0`;
      } else if (placement.startsWith('left')) {
        element.padding = `0 ${length}px 0 0`;
      } else if (placement.startsWith('right')) {
        element.padding = `0 0 0 ${length}px`;
      }
    }

    if (status === STATUS.CLOSING) {
      element = { ...element, ...floaterClosing };
    }

    if (status === STATUS.OPENING || status === STATUS.OPEN) {
      element = { ...element, ...floaterOpening };
    }

    if (placement === 'center') {
      element = { ...element, ...floaterCentered };
    }

    if (component) {
      element = { ...element, ...floaterWithComponent };
    }

    return element;
  }, [component, hideArrow, placement, status, styles]);

  const shouldRender = ['closing', 'open', 'opening', 'render'].includes(status);

  const output: PlainObject<ReactNode> = {};
  const classes = ['__floater'];
  const baseProps = { role: 'tooltip' };

  if (component) {
    const componentProps = { closeFn, id, ...baseProps };

    output.content = isValidElement(component)
      ? cloneElement(component, componentProps)
      : component(componentProps);
  } else {
    output.content = <Container {...props} content={content} />;
  }

  if (status === STATUS.OPEN) {
    classes.push('__floater__open');
  }

  if (!hideArrow) {
    output.arrow = <Arrow {...props} />;
  }

  if (!shouldRender) {
    return null;
  }

  return (
    <div ref={floaterRef} id={id} style={{ zIndex: styles.options.zIndex }}>
      <div className={classes.join(' ')} style={style} {...baseProps}>
        <div className="__floater__body">
          {output.content}
          {output.arrow}
        </div>
      </div>
    </div>
  );
}

export default memo(Floater);
