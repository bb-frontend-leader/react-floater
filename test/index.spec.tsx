/* eslint-disable testing-library/no-manual-cleanup */
import { ReactNode } from 'react';
import {
  act,
  cleanup,
  configure,
  fireEvent,
  render,
  RenderResult,
  screen,
} from '@testing-library/react';
import { MockInstance } from 'vitest';

import ReactFloater from '../src';
import { portalId } from '../src/modules/helpers';
import { Props } from '../src/types';

import { Button, Floaters, Styled } from './__fixtures__/components';

configure({
  testIdAttribute: 'id',
});

vi.useFakeTimers();

const mockCallback = vi.fn();
const mockGetPopper = vi.fn(() => ({ instance: {} }));

const id = 'test';
const content = 'Hello! This is my content!';

const props: Props = {
  id,
  content,
  getPopper: mockGetPopper,
};

function setup(ownProps = props, children: ReactNode = 'Places') {
  return render(<ReactFloater {...ownProps}>{children}</ReactFloater>);
}

describe('ReactFloater', () => {
  let view: RenderResult;

  const unmountView = () => {
    view?.unmount();
    cleanup();
  };

  afterEach(() => {
    mockCallback.mockClear();
  });

  describe('basic usage', () => {
    afterAll(() => {
      mockGetPopper.mockClear();
      unmountView();
    });

    it('should render the element', () => {
      view = setup(props, 'Places');

      expect(screen.getByTestId(`${id}-wrapper`)).toBeInTheDocument();
    });

    it('should render the portal, popper and floater', async () => {
      fireEvent.click(screen.getByTestId(`${id}-wrapper`));
      const portal = document.getElementById('react-floater-portal');
      const popper = portal?.firstChild;

      expect(portal).toBeInTheDocument();
      expect(portal).toHaveStyle({ zIndex: 100 });
      expect(popper).toHaveStyle({ zIndex: 100 });

      expect(screen.getByTestId('test')).toHaveClass('__floater');
      expect(screen.getByTestId('test')).not.toHaveClass('__floater__open');
    });

    it('should have called getPopper', () => {
      expect(mockGetPopper).toHaveBeenCalledWith(expect.any(Object), 'floater');
    });

    it('should show the floater after the transition ends', () => {
      fireEvent.transitionEnd(screen.getByTestId('test'));
      expect(screen.getByTestId('test')).toHaveClass('__floater__open');
    });

    it('should hide the floater and remove the popper', async () => {
      fireEvent.click(screen.getByTestId(`${id}-wrapper`));

      fireEvent.transitionEnd(screen.getByTestId('test'));

      const portal = document.getElementById('react-floater-portal');
      const popper = portal?.firstChild;

      expect(portal).toBeInTheDocument();
      expect(popper).toBeNull();

      expect(screen.getByTestId('react-floater-portal')).toBeEmptyDOMElement();
    });
  });

  describe('with multiple children', () => {
    afterAll(unmountView);

    it('should render the element', () => {
      view = setup(
        props,
        <>
          <span>Hello</span> <span>world</span>
        </>,
      );

      expect(screen.getByTestId(`${id}-wrapper`)).toBeInTheDocument();
    });

    it('should render the floater', async () => {
      fireEvent.click(screen.getByTestId(`${id}-wrapper`));

      expect(screen.getByTestId('test')).toBeInTheDocument();
    });

    it('should hide the floater', async () => {
      fireEvent.click(screen.getByTestId(`${id}-wrapper`));

      fireEvent.transitionEnd(screen.getByTestId('test'));

      expect(screen.getByTestId('react-floater-portal')).toBeEmptyDOMElement();
    });
  });

  describe('with multiple instances', () => {
    afterAll(unmountView);

    it('should render the elements', () => {
      view = render(<Floaters />);

      expect(screen.getByTestId('president-wrapper')).toBeInTheDocument();
      expect(screen.getByTestId('republic-wrapper')).toBeInTheDocument();
    });

    it('should render the floaters', async () => {
      fireEvent.click(screen.getByTestId('president-wrapper'));
      fireEvent.click(screen.getByTestId('republic-wrapper'));

      expect(screen.getByTestId('president')).toHaveTextContent('It was that bearded guy!');
      expect(screen.getByTestId('republic')).toHaveTextContent('You know what I mean');
    });

    it('should hide the floaters', async () => {
      fireEvent.click(screen.getByTestId('president-wrapper'));
      fireEvent.click(screen.getByTestId('republic-wrapper'));

      fireEvent.transitionEnd(screen.getByTestId('president'));
      fireEvent.transitionEnd(screen.getByTestId('republic'));

      expect(screen.getByTestId('react-floater-portal')).toBeEmptyDOMElement();
    });
  });

  describe('with `portalElement`', () => {
    let element: HTMLDivElement;

    beforeAll(() => {
      element = document.createElement('div');
      element.id = 'floaters';

      document.body.appendChild(element);
    });

    afterAll(() => {
      unmountView();
      document.body.removeChild(element);
    });

    it('should render in the "portalElement"', async () => {
      view = render(<Floaters portalElement={element} />);

      expect(screen.queryByTestId(portalId)).not.toBeInTheDocument();
      expect(screen.getByTestId('floaters')).toBeInTheDocument();

      fireEvent.click(screen.getByTestId('president-wrapper'));
      fireEvent.click(screen.getByTestId('republic-wrapper'));

      expect(screen.getByTestId('president')).toHaveTextContent('It was that bearded guy!');
      expect(screen.getByTestId('republic')).toHaveTextContent('You know what I mean');
    });
  });

  describe('with `autoOpen`', () => {
    afterAll(unmountView);

    it('should have rendered the Floater initially open', async () => {
      view = setup({
        ...props,
        autoOpen: true,
      });

      await act(async () => {
        vi.runOnlyPendingTimers();
      });

      fireEvent.transitionEnd(screen.getByTestId('test'));
      expect(screen.getByTestId('test')).toHaveClass('__floater__open');
    });

    it('should hide the floater', async () => {
      fireEvent.click(screen.getByTestId(`${id}-wrapper`));

      fireEvent.transitionEnd(screen.getByTestId('test'));

      expect(screen.getByTestId('react-floater-portal')).toBeEmptyDOMElement();
    });

    it('should show the floater again', async () => {
      fireEvent.click(screen.getByTestId(`${id}-wrapper`));

      await act(async () => {
        vi.runOnlyPendingTimers();
      });

      fireEvent.transitionEnd(screen.getByTestId('test'));

      expect(screen.getByTestId('test')).toHaveClass('__floater__open');
    });
  });

  describe('with `callback`', () => {
    afterAll(unmountView);

    it('should call the callback function on open', async () => {
      view = setup({
        ...props,
        callback: mockCallback,
      });

      fireEvent.click(screen.getByTestId(`${id}-wrapper`));

      await act(async () => {
        vi.runOnlyPendingTimers();
      });

      fireEvent.transitionEnd(screen.getByTestId('test'));

      expect(mockCallback).toHaveBeenCalledWith('open', {
        autoOpen: false,
        callback: mockCallback,
        children: 'Places',
        content: 'Hello! This is my content!',
        debug: false,
        disableFlip: false,
        disableHoverToClick: false,
        event: 'click',
        eventDelay: 0.4,
        getPopper: mockGetPopper,
        hideArrow: false,
        id,
        offset: 15,
        placement: 'bottom',
        showCloseButton: false,
        styles: {},
        target: null,
        wrapperOptions: { position: false },
      });
    });

    it('should call the callback function on close', async () => {
      fireEvent.click(screen.getByTestId(`${id}-wrapper`));

      fireEvent.transitionEnd(screen.getByTestId('test'));

      expect(mockCallback).toHaveBeenCalledWith('close', {
        autoOpen: false,
        callback: mockCallback,
        children: 'Places',
        content: 'Hello! This is my content!',
        debug: false,
        disableFlip: false,
        disableHoverToClick: false,
        event: 'click',
        eventDelay: 0.4,
        getPopper: mockGetPopper,
        hideArrow: false,
        id,
        offset: 15,
        placement: 'bottom',
        showCloseButton: false,
        styles: {},
        target: null,
        wrapperOptions: { position: false },
      });
    });
  });

  describe('with `debug`', () => {
    let consoleGroupCollapsed: MockInstance;
    let consoleLog: MockInstance;

    beforeAll(() => {
      consoleGroupCollapsed = vi
        .spyOn(console, 'groupCollapsed')
        .mockImplementation(() => undefined);
      consoleLog = vi.spyOn(console, 'log').mockImplementation(() => undefined);
    });

    afterAll(() => {
      consoleGroupCollapsed.mockRestore();
      consoleLog.mockRestore();
      unmountView();
    });

    it('should be able to show and hide the floater', async () => {
      view = setup({
        ...props,
        debug: true,
      });

      expect(consoleGroupCollapsed).toHaveBeenCalledWith(
        '%creact-floater: init',
        'color: #9b00ff; font-weight: bold; font-size: 12px;',
      );

      expect(consoleLog).toHaveBeenCalledWith(
        expect.objectContaining({
          floater: null,
          hasChildren: true,
          hasTarget: false,
          isControlled: false,
          positionWrapper: false,
          target: expect.anything(),
        }),
      );

      fireEvent.click(screen.getByTestId(`${id}-wrapper`));

      expect(consoleGroupCollapsed).toHaveBeenCalledWith(
        '%creact-floater: click',
        'color: #9b00ff; font-weight: bold; font-size: 12px;',
      );
      expect(consoleLog).toHaveBeenCalledWith({ event: 'click', status: 'opening' });

      await act(async () => {
        vi.runOnlyPendingTimers();
      });

      fireEvent.transitionEnd(screen.getByTestId('test'));

      expect(screen.getByTestId('test')).toHaveClass('__floater__open');

      expect(consoleGroupCollapsed).toHaveBeenCalledTimes(2);
      expect(consoleLog).toHaveBeenCalledTimes(2);
    });
  });

  describe('with `event` hover', () => {
    afterAll(unmountView);

    it('should be able to show the floater', async () => {
      view = setup({
        ...props,
        event: 'hover',
        disableHoverToClick: true,
      });

      fireEvent.mouseEnter(screen.getByTestId(`${id}-wrapper`));

      await act(async () => {
        vi.runOnlyPendingTimers();
      });

      fireEvent.transitionEnd(screen.getByTestId('test'));

      expect(screen.getByTestId('test')).toHaveClass('__floater__open');
    });

    it('should close itself after `eventDelay`', async () => {
      fireEvent.mouseLeave(screen.getByTestId(`${id}-wrapper`));

      act(() => {
        vi.advanceTimersByTime(1000);
      });

      expect(screen.getByTestId('test')).not.toHaveClass('__floater__open');

      fireEvent.transitionEnd(screen.getByTestId('test'));

      expect(screen.getByTestId('react-floater-portal')).toBeEmptyDOMElement();
    });
  });

  describe('with `event` hover and `eventDelay` set to 0', () => {
    afterAll(unmountView);

    it('should be able to show the floater', async () => {
      view = setup({
        ...props,
        event: 'hover',
        eventDelay: 0,
      });

      fireEvent.mouseEnter(screen.getByTestId(`${id}-wrapper`));

      await act(async () => {
        vi.runOnlyPendingTimers();
      });

      fireEvent.transitionEnd(screen.getByTestId('test'));

      expect(screen.getByTestId('test')).toHaveClass('__floater__open');
    });

    it('should have close itself immediately', async () => {
      fireEvent.mouseLeave(screen.getByTestId(`${id}-wrapper`));

      fireEvent.transitionEnd(screen.getByTestId('test'));

      expect(screen.getByTestId('react-floater-portal')).toBeEmptyDOMElement();
    });
  });

  describe('with `title`, `footer` and `closeBtn`', () => {
    afterAll(unmountView);

    it('should render the floater', async () => {
      view = setup({
        ...props,
        footer: 'Footer',
        showCloseButton: true,
        title: 'Title',
      });

      fireEvent.click(screen.getByTestId(`${id}-wrapper`));

      await act(async () => {
        vi.runOnlyPendingTimers();
      });

      fireEvent.transitionEnd(screen.getByTestId('test'));

      expect(screen.getByTestId('test')).toMatchSnapshot();
    });
  });

  describe('with `open`', () => {
    const localProps = {
      ...props,
      open: false,
    };

    afterAll(unmountView);

    it('should not be able to show the floater with click', async () => {
      view = setup(localProps);

      fireEvent.click(screen.getByTestId(`${id}-wrapper`));

      expect(screen.queryByTestId('test')).not.toBeInTheDocument();
    });

    it('should not be able to show the floater with hover', async () => {
      view = setup({ ...localProps, event: 'hover' });

      fireEvent.mouseEnter(screen.getAllByTestId(`${id}-wrapper`)[0]);

      expect(screen.queryByTestId('test')).not.toBeInTheDocument();
    });

    it('should show the floater when `open` is true', async () => {
      view.rerender(
        <ReactFloater {...localProps} open>
          {content}
        </ReactFloater>,
      );

      await act(async () => {
        vi.runOnlyPendingTimers();
      });

      fireEvent.transitionEnd(screen.getByTestId('test'));

      expect(screen.getByTestId('test')).toHaveClass('__floater__open');
    });

    it('should close the floater when `open` is false', async () => {
      view.rerender(
        <ReactFloater {...localProps} open={false}>
          {content}
        </ReactFloater>,
      );

      fireEvent.transitionEnd(screen.getByTestId('test'));

      expect(screen.queryByTestId('test')).not.toBeInTheDocument();
    });
  });

  describe('with `component` as function', () => {
    afterAll(unmountView);

    it('should show the floater with click', async () => {
      view = setup(
        {
          ...props,
          content: undefined,
          component: Styled,
        },
        <Button />,
      );

      fireEvent.click(screen.getByTestId('button'));

      await act(async () => {
        vi.runOnlyPendingTimers();
      });

      expect(screen.getByTestId('test')).toMatchSnapshot();
    });

    it('should close the floater with `closeFn` prop', async () => {
      fireEvent.click(screen.getByRole('button', { name: /close/i }));

      fireEvent.transitionEnd(screen.getByTestId('test'));

      expect(screen.getByTestId('react-floater-portal')).toBeEmptyDOMElement();
    });
  });

  describe('with `component` as element', () => {
    afterAll(unmountView);

    it('should show the floater with click', async () => {
      view = setup(
        {
          ...props,
          content: undefined,
          component: <Styled />,
          styles: {
            options: {
              zIndex: 1000,
            },
          },
        },
        <span>Places</span>,
      );

      fireEvent.click(screen.getByTestId(`${id}-wrapper`));

      await act(async () => {
        vi.runOnlyPendingTimers();
      });

      expect(screen.getByTestId('test')).toMatchSnapshot();
    });

    it('should close the floater with `closeFn` prop', async () => {
      fireEvent.click(screen.getByRole('button', { name: /close/i }));

      fireEvent.transitionEnd(screen.getByTestId('test'));

      expect(screen.getByTestId('react-floater-portal')).toBeEmptyDOMElement();
    });
  });

  describe('with `placement` top', () => {
    afterAll(unmountView);

    it('should show the floater with click', async () => {
      view = setup({
        ...props,
        placement: 'top',
      });

      fireEvent.click(screen.getByTestId(`${id}-wrapper`));

      await act(async () => {
        vi.runOnlyPendingTimers();
      });

      expect(screen.getByTestId('test')).toMatchSnapshot();
    });
  });

  describe('with `placement` left', () => {
    afterAll(unmountView);

    it('should show the floater with click', async () => {
      view = setup({
        ...props,
        placement: 'left',
      });

      fireEvent.click(screen.getByTestId(`${id}-wrapper`));

      await act(async () => {
        vi.runOnlyPendingTimers();
      });

      expect(screen.getByTestId('test')).toMatchSnapshot();
    });
  });

  describe('with `placement` right', () => {
    afterAll(unmountView);

    it('should show the floater with click', async () => {
      view = setup({
        ...props,
        placement: 'right',
      });

      fireEvent.click(screen.getByTestId(`${id}-wrapper`));

      await act(async () => {
        vi.runOnlyPendingTimers();
      });

      expect(screen.getByTestId('test')).toMatchSnapshot();
    });
  });

  describe('with `placement` center', () => {
    afterAll(unmountView);

    it('should show the floater with click', async () => {
      view = setup({
        ...props,
        placement: 'center',
      });
      fireEvent.click(screen.getByTestId(`${id}-wrapper`));

      await act(async () => {
        vi.runOnlyPendingTimers();
      });

      expect(screen.getByTestId(id)).toMatchSnapshot();
    });
  });

  describe('with `wrapperOptions`', () => {
    afterAll(unmountView);

    it('should render properly', async () => {
      view = render(
        <>
          <span className="external" id="external">
            external
          </span>
          <ReactFloater
            content={
              <div style={{ fontSize: 32 }}>Yeah, this is how we use to look back in the day!</div>
            }
            disableFlip
            id={id}
            placement="top"
            target=".external"
            wrapperOptions={{
              offset: -20,
              placement: 'bottom',
              position: true,
            }}
          >
            <span id="beacon">Beacon</span>
          </ReactFloater>
        </>,
      );

      fireEvent.click(screen.getByTestId('test-wrapper'));

      expect(screen.getByTestId(id)).toHaveTextContent(
        'Yeah, this is how we use to look back in the day!',
      );
    });
  });
});
