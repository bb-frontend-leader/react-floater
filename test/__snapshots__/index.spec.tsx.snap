// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ReactFloater > with \`arrow\` > should render a custom arrow floater with click 1`] = `
<div
  data-popper-escaped=""
  data-popper-placement="bottom"
  data-popper-reference-hidden=""
  id="test"
  style="z-index: 100; position: absolute; left: 0px; top: 0px; margin: 0px; right: auto; bottom: auto; transform: translate(0px, 15px);"
>
  <div
    class="__floater"
    role="tooltip"
    style="display: inline-block; filter: drop-shadow(0 0 3px rgba(0, 0, 0, 0.3)); max-width: 300px; opacity: 1; position: relative; transition: opacity 0.3s; visibility: visible; padding: 80px 0px 0px;"
  >
    <div
      class="__floater__body"
    >
      <div
        class="__floater__container"
        style="background-color: rgb(255, 255, 255); color: rgb(102, 102, 102); min-height: 60px; min-width: 200px; padding: 20px; position: relative; z-index: 10;"
      >
        <div
          class="__floater__content"
          style="font-size: 15px;"
        >
          Hello! This is my content!
        </div>
      </div>
      <span
        style="color: rgb(107, 162, 255); height: 80px; width: 10px; position: absolute; left: 0px; transform: translate(8px, 0px); top: 0px;"
      >
        <svg
          height="100px"
          version="1.1"
          viewBox="0 0 10 100"
          width="10px"
          xmlns="http://www.w3.org/2000/svg"
        >
          <g>
            <path
              d="M5.19249228e-07,5.16768813 C0.000571745188,2.3180242 2.3238984,6.52762368e-06 5.17356233,6.52762368e-06 C7.4961469,-0.00369898343 9.46628897,1.57062447 10.1275925,3.67813093 C11.5918888,8.26905035 5.18127186,100.000007 5.18127186,100.000007 L1.9635443,41.3615605 C0.907405183,22.0686414 -0.000792731847,5.40502691 5.19249228e-07,5.16768813 Z M5.18507488,2.58554059 C3.75638551,2.5874823 2.59249512,3.74291582 2.59443184,5.17160518 C2.59160462,6.52075954 3.7928988,13.3574741 4.95859452,14.5045202 C6.56709198,13.5758538 7.69207267,4.78979882 7.69207267,4.78979882 C7.50203478,3.55199808 6.48229278,2.58749911 5.18507488,2.58554059 Z"
              fill="currentColor"
            />
          </g>
        </svg>
      </span>
    </div>
  </div>
</div>
`;

exports[`ReactFloater > with \`component\` as element > should show the floater with click 1`] = `
<div
  data-popper-escaped=""
  data-popper-placement="bottom"
  data-popper-reference-hidden=""
  id="test"
  style="z-index: 1000; position: absolute; left: 0px; top: 0px; margin: 0px; right: auto; bottom: auto; transform: translate(0px, 15px);"
>
  <div
    class="__floater"
    role="tooltip"
    style="display: inline-block; filter: drop-shadow(0 0 3px rgba(0, 0, 0, 0.3)); max-width: 100%; opacity: 1; position: relative; transition: opacity 0.3s; visibility: visible; z-index: 1000; padding: 16px 0px 0px;"
  >
    <div
      class="__floater__body"
    >
      <div
        class="css-lgzz4i"
      >
        <div>
          Styled
        </div>
        <button
          type="button"
        >
          close
        </button>
      </div>
      <span
        class="__floater__arrow"
        style="display: inline-flex; position: absolute; left: 0px; transform: translate(8px, 0px); top: 0px;"
      >
        <svg
          height="16"
          version="1.1"
          width="32"
          xmlns="http://www.w3.org/2000/svg"
        >
          <polygon
            fill="#fff"
            points="32,16 16,0 0,16"
          />
        </svg>
      </span>
    </div>
  </div>
</div>
`;

exports[`ReactFloater > with \`component\` as function > should show the floater with click 1`] = `
<div
  data-popper-escaped=""
  data-popper-placement="bottom"
  data-popper-reference-hidden=""
  id="test"
  style="z-index: 100; position: absolute; left: 0px; top: 0px; margin: 0px; right: auto; bottom: auto; transform: translate(0px, 15px);"
>
  <div
    class="__floater"
    role="tooltip"
    style="display: inline-block; filter: drop-shadow(0 0 3px rgba(0, 0, 0, 0.3)); max-width: 100%; opacity: 1; position: relative; transition: opacity 0.3s; visibility: visible; padding: 16px 0px 0px;"
  >
    <div
      class="__floater__body"
    >
      <div
        class="css-lgzz4i"
      >
        <div>
          Styled
        </div>
        <button
          type="button"
        >
          close
        </button>
      </div>
      <span
        class="__floater__arrow"
        style="display: inline-flex; position: absolute; left: 0px; transform: translate(8px, 0px); top: 0px;"
      >
        <svg
          height="16"
          version="1.1"
          width="32"
          xmlns="http://www.w3.org/2000/svg"
        >
          <polygon
            fill="#fff"
            points="32,16 16,0 0,16"
          />
        </svg>
      </span>
    </div>
  </div>
</div>
`;

exports[`ReactFloater > with \`placement\` center > should show the floater with click 1`] = `
<div
  id="test"
  style="z-index: 100;"
>
  <div
    class="__floater"
    role="tooltip"
    style="display: inline-block; filter: drop-shadow(0 0 3px rgba(0, 0, 0, 0.3)); max-width: 300px; opacity: 1; position: fixed; transition: opacity 0.3s; visibility: visible; left: 50%; top: 50%; transform: translate(-50%, -50%);"
  >
    <div
      class="__floater__body"
    >
      <div
        class="__floater__container"
        style="background-color: rgb(255, 255, 255); color: rgb(102, 102, 102); min-height: 60px; min-width: 200px; padding: 20px; position: relative; z-index: 10;"
      >
        <div
          class="__floater__content"
          style="font-size: 15px;"
        >
          Hello! This is my content!
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`ReactFloater > with \`placement\` left > should show the floater with click 1`] = `
<div
  data-popper-escaped=""
  data-popper-placement="left"
  data-popper-reference-hidden=""
  id="test"
  style="z-index: 100; position: absolute; left: auto; top: 0px; margin: 0px; right: 0px; bottom: auto; transform: translate(-15px, 0px);"
>
  <div
    class="__floater"
    role="tooltip"
    style="display: inline-block; filter: drop-shadow(0 0 3px rgba(0, 0, 0, 0.3)); max-width: 300px; opacity: 1; position: relative; transition: opacity 0.3s; visibility: visible; padding: 0px 16px 0px 0px;"
  >
    <div
      class="__floater__body"
    >
      <div
        class="__floater__container"
        style="background-color: rgb(255, 255, 255); color: rgb(102, 102, 102); min-height: 60px; min-width: 200px; padding: 20px; position: relative; z-index: 10;"
      >
        <div
          class="__floater__content"
          style="font-size: 15px;"
        >
          Hello! This is my content!
        </div>
      </div>
      <span
        class="__floater__arrow"
        style="display: inline-flex; position: absolute; top: 0px; transform: translate(0px, 8px); right: 0px;"
      >
        <svg
          height="32"
          version="1.1"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <polygon
            fill="#fff"
            points="0,0 16,16 0,32"
          />
        </svg>
      </span>
    </div>
  </div>
</div>
`;

exports[`ReactFloater > with \`placement\` right > should show the floater with click 1`] = `
<div
  data-popper-escaped=""
  data-popper-placement="right"
  data-popper-reference-hidden=""
  id="test"
  style="z-index: 100; position: absolute; left: 0px; top: 0px; margin: 0px; right: auto; bottom: auto; transform: translate(15px, 0px);"
>
  <div
    class="__floater"
    role="tooltip"
    style="display: inline-block; filter: drop-shadow(0 0 3px rgba(0, 0, 0, 0.3)); max-width: 300px; opacity: 1; position: relative; transition: opacity 0.3s; visibility: visible; padding: 0px 0px 0px 16px;"
  >
    <div
      class="__floater__body"
    >
      <div
        class="__floater__container"
        style="background-color: rgb(255, 255, 255); color: rgb(102, 102, 102); min-height: 60px; min-width: 200px; padding: 20px; position: relative; z-index: 10;"
      >
        <div
          class="__floater__content"
          style="font-size: 15px;"
        >
          Hello! This is my content!
        </div>
      </div>
      <span
        class="__floater__arrow"
        style="display: inline-flex; position: absolute; top: 0px; transform: translate(0px, 8px); left: 0px;"
      >
        <svg
          height="32"
          version="1.1"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <polygon
            fill="#fff"
            points="16,32 16,0 0,16"
          />
        </svg>
      </span>
    </div>
  </div>
</div>
`;

exports[`ReactFloater > with \`placement\` top > should show the floater with click 1`] = `
<div
  data-popper-escaped=""
  data-popper-placement="top"
  data-popper-reference-hidden=""
  id="test"
  style="z-index: 100; position: absolute; left: 0px; top: auto; margin: 0px; right: auto; bottom: 0px; transform: translate(0px, -15px);"
>
  <div
    class="__floater"
    role="tooltip"
    style="display: inline-block; filter: drop-shadow(0 0 3px rgba(0, 0, 0, 0.3)); max-width: 300px; opacity: 1; position: relative; transition: opacity 0.3s; visibility: visible; padding: 0px 0px 16px;"
  >
    <div
      class="__floater__body"
    >
      <div
        class="__floater__container"
        style="background-color: rgb(255, 255, 255); color: rgb(102, 102, 102); min-height: 60px; min-width: 200px; padding: 20px; position: relative; z-index: 10;"
      >
        <div
          class="__floater__content"
          style="font-size: 15px;"
        >
          Hello! This is my content!
        </div>
      </div>
      <span
        class="__floater__arrow"
        style="display: inline-flex; position: absolute; left: 0px; transform: translate(8px, 0px); bottom: 0px;"
      >
        <svg
          height="16"
          version="1.1"
          width="32"
          xmlns="http://www.w3.org/2000/svg"
        >
          <polygon
            fill="#fff"
            points="0,0 16,16 32,0"
          />
        </svg>
      </span>
    </div>
  </div>
</div>
`;

exports[`ReactFloater > with \`title\`, \`footer\` and \`closeBtn\` > should render the floater 1`] = `
<div
  data-popper-escaped=""
  data-popper-placement="bottom"
  data-popper-reference-hidden=""
  id="test"
  style="z-index: 100; position: absolute; left: 0px; top: 0px; margin: 0px; right: auto; bottom: auto; transform: translate(0px, 15px);"
>
  <div
    class="__floater __floater__open"
    role="tooltip"
    style="display: inline-block; filter: drop-shadow(0 0 3px rgba(0, 0, 0, 0.3)); max-width: 300px; opacity: 1; position: relative; transition: opacity 0.3s; visibility: visible; padding: 16px 0px 0px;"
  >
    <div
      class="__floater__body"
    >
      <div
        class="__floater__container"
        style="background-color: rgb(255, 255, 255); color: rgb(102, 102, 102); min-height: 60px; min-width: 200px; padding: 20px; position: relative; z-index: 10;"
      >
        <button
          aria-label="close"
          style="background-color: transparent; border: 0px; border-radius: 0; font-size: 0px; outline: none; padding: 10px; position: absolute; right: 0px; top: 0px;"
          type="button"
        >
          <svg
            height="15px"
            preserveAspectRatio="xMidYMid"
            version="1.1"
            viewBox="0 0 18 18"
            width="15px"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g>
              <path
                d="M8.13911129,9.00268191 L0.171521827,17.0258467 C-0.0498027049,17.248715 -0.0498027049,17.6098394 0.171521827,17.8327545 C0.28204354,17.9443526 0.427188206,17.9998706 0.572051765,17.9998706 C0.71714958,17.9998706 0.862013139,17.9443526 0.972581703,17.8327545 L9.0000937,9.74924618 L17.0276057,17.8327545 C17.1384085,17.9443526 17.2832721,17.9998706 17.4281356,17.9998706 C17.5729992,17.9998706 17.718097,17.9443526 17.8286656,17.8327545 C18.0499901,17.6098862 18.0499901,17.2487618 17.8286656,17.0258467 L9.86135722,9.00268191 L17.8340066,0.973848225 C18.0553311,0.750979934 18.0553311,0.389855532 17.8340066,0.16694039 C17.6126821,-0.0556467968 17.254037,-0.0556467968 17.0329467,0.16694039 L9.00042166,8.25611765 L0.967006424,0.167268345 C0.745681892,-0.0553188426 0.387317931,-0.0553188426 0.165993399,0.167268345 C-0.0553311331,0.390136635 -0.0553311331,0.751261038 0.165993399,0.974176179 L8.13920499,9.00268191 L8.13911129,9.00268191 Z"
                fill="#555"
              />
            </g>
          </svg>
        </button>
        <div
          class="__floater__title"
          style="border-bottom: 1px solid #555; color: rgb(85, 85, 85); font-size: 18px; margin-bottom: 5px; padding-bottom: 6px; padding-right: 18px;"
        >
          Title
        </div>
        <div
          class="__floater__content"
          style="font-size: 15px;"
        >
          Hello! This is my content!
        </div>
        <div
          class="__floater__footer"
          style="border-top: 1px solid #ccc; font-size: 13px; margin-top: 10px; padding-top: 5px;"
        >
          Footer
        </div>
      </div>
      <span
        class="__floater__arrow"
        style="display: inline-flex; position: absolute; left: 0px; transform: translate(8px, 0px); top: 0px;"
      >
        <svg
          height="16"
          version="1.1"
          width="32"
          xmlns="http://www.w3.org/2000/svg"
        >
          <polygon
            fill="#fff"
            points="32,16 16,0 0,16"
          />
        </svg>
      </span>
    </div>
  </div>
</div>
`;
